cpp/src/arrow/util/decimal-benchmark.cc

```diff
@@ -45,53 +45,114 @@ static void FromString(benchmark::State& state) {  // NOLINT non-const reference
 }
 
 static void BinaryCompareOp(benchmark::State& state) {  // NOLINT non-const reference
-  BasicDecimal128 d1(546, 123), d2(123, 456);
-  while (state.KeepRunning()) {
-    benchmark::DoNotOptimize(d1 == d2);
-    benchmark::DoNotOptimize(d1 <= d2);
-    benchmark::DoNotOptimize(d1 >= d2);
-    benchmark::DoNotOptimize(d1 >= d1);
+  std::vector<BasicDecimal128> v1, v2;
+  for (int x = 0; x < 100; x++) {
+    v1.emplace_back(100 + x, 100 + x);
+    v2.emplace_back(200 + x, 200 + x);
+  }
+  for (auto _ : state) {
+    for (int x = 0; x < 25; x += 4) {
+      benchmark::DoNotOptimize(v1[x] == v2[x]);
+      benchmark::DoNotOptimize(v1[x + 1] <= v2[x + 1]);
+      benchmark::DoNotOptimize(v1[x + 2] >= v2[x + 2]);
+      benchmark::DoNotOptimize(v1[x + 3] >= v1[x + 3]);
+    }
+  }
+}
+
+static void BinaryCompareOpConstant(
+    benchmark::State& state) {  // NOLINT non-const reference
+  std::vector<BasicDecimal128> v1, v2;
+  for (int x = 0; x < 100; x++) {
+    v1.emplace_back(100 + x, 100 + x);
+    v2.emplace_back(200 + x, 200 + x);
+  }
+  for (auto _ : state) {
+    for (int x = 0; x < 25; x += 4) {
+      v2[x] = v1[x] == v1[0];
```

Comment:
Is the combination of `=` and `==` intended here? You're assigning a bool to a BasicDecimal?